pkg/kubelet/images/pullmanager/mem_pullrecords.go

```diff
@@ -0,0 +1,186 @@
+	pulledRecordsMutex *sync.RWMutex
+	pulledRecords      *LRUCache[string, kubeletconfiginternal.ImagePulledRecord]
+}
+
+func NewCachedPullRecordsAccessor(delegate PullRecordsAccessor) *cachedPullRecordsAccessor {
+	// TODO: handle errors from records listing here and below
+	coldIntents, _ := delegate.ListImagePullIntents()
+	warmIntents := NewLRUCache[string, kubeletconfiginternal.ImagePullIntent](50)
+	for _, intent := range coldIntents[:min(50, len(coldIntents))] {
+		warmIntents.Set(intent.Image, intent)
+	}
+
+	coldPulledRecords, _ := delegate.ListImagePulledRecords()
+	warmPulledRecords := NewLRUCache[string, kubeletconfiginternal.ImagePulledRecord](100)
+	for _, pulledRecord := range coldPulledRecords[:min(100, len(coldPulledRecords))] {
+		warmPulledRecords.Set(pulledRecord.ImageRef, pulledRecord)
+	}
+
+	return &cachedPullRecordsAccessor{
+		delegate: delegate,
+
+		intentsMutex:       &sync.RWMutex{},
+		intents:            warmIntents,
+		pulledRecordsMutex: &sync.RWMutex{},
+		pulledRecords:      warmPulledRecords,
+	}
+}
+
+func (c *cachedPullRecordsAccessor) ListImagePullIntents() ([]*kubeletconfiginternal.ImagePullIntent, error) {
+	//TODO: maybe this should be protected by a read lock
+	c.intentsMutex.RLock()
+	defer c.intentsMutex.RUnlock()
+	return c.delegate.ListImagePullIntents()
+}
+
+func (c *cachedPullRecordsAccessor) ImagePullIntentExists(image string) (bool, error) {
+	if _, exists := c.intents.Get(image); exists {
+		return true, nil
+	}
+
+	c.intentsMutex.RLock()
+	defer c.intentsMutex.RUnlock()
+
+	exists, err := c.delegate.ImagePullIntentExists(image)
+	if err == nil && exists {
+		c.intents.Set(image, &kubeletconfiginternal.ImagePullIntent{
+			Image: image,
+		})
+	}
+	return exists, err
+}
+
+func (c *cachedPullRecordsAccessor) WriteImagePullIntent(image string) error {
+	c.intentsMutex.Lock()
+	defer c.intentsMutex.Unlock()
+
+	if err := c.delegate.WriteImagePullIntent(image); err != nil {
+		return err
+	}
+
+	c.intents.Set(image, &kubeletconfiginternal.ImagePullIntent{
+		Image: image,
+	})
+	return nil
+}
+
+func (c *cachedPullRecordsAccessor) DeleteImagePullIntent(image string) error {
+	c.intentsMutex.Lock()
+	defer c.intentsMutex.Unlock()
+
+	if err := c.delegate.DeleteImagePullIntent(image); err != nil {
+		return err
+	}
+	c.intents.Delete(image)
+	return nil
+}
+
+func (c *cachedPullRecordsAccessor) ListImagePulledRecords() ([]*kubeletconfiginternal.ImagePulledRecord, error) {
+	//TODO: maybe this should be protected by a read lock
+	c.pulledRecordsMutex.RLock()
+	defer c.pulledRecordsMutex.RUnlock()
+	return c.delegate.ListImagePulledRecords()
+}
+
+func (c *cachedPullRecordsAccessor) GetImagePulledRecord(imageRef string) (*kubeletconfiginternal.ImagePulledRecord, bool, error) {
+	pulledRecord, exists := c.pulledRecords.Get(imageRef)
+	if exists {
+		return pulledRecord, true, nil
+	}
+
+	c.pulledRecordsMutex.RLock()
+	defer c.pulledRecordsMutex.RUnlock()
```

Comment:
similar comment, it's odd to check `c.pulledRecords` locally, then take a global **read** lock, then **write** to `c.pulledRecords`

what is this attempting to protect by locking?